angular/projects/admin-nrpti/src/app/records/orders/order-lng-detail/order-lng-detail.component.ts

```diff
@@ -0,0 +1,96 @@
+import { Component, OnInit, OnDestroy } from '@angular/core';
+import { takeUntil } from 'rxjs/operators';
+import { OrderLNG } from '../../../../../../common/src/app/models/lng/order-lng';
+import { Subject } from 'rxjs';
+import { ActivatedRoute, Router } from '@angular/router';
+import { RecordComponent } from '../../utils/record-component';
+import { DatePipe } from '@angular/common';
+import { FactoryService } from '../../../services/factory.service';
+
+@Component({
+  selector: 'app-order-lng-detail',
+  templateUrl: './order-lng-detail.component.html',
+  styleUrls: ['./order-lng-detail.component.scss']
+})
+export class OrderLNGDetailComponent extends RecordComponent implements OnInit, OnDestroy {
+  private ngUnsubscribe: Subject<boolean> = new Subject<boolean>();
+
+  public isLoading = true;
+  public isPublishing = false;
+  public isUnPublishing = false;
+
+  constructor(
+    public route: ActivatedRoute,
+    public router: Router,
+    public factoryService: FactoryService,
+    public datePipe: DatePipe
+  ) {
+    super();
+  }
+
+  ngOnInit() {
+    if (this.data) {
+      this.isLoading = false;
+      return;
+    }
+
+    this.route.data.pipe(takeUntil(this.ngUnsubscribe)).subscribe((res: any) => {
+      if (!res || !res.records) {
+        alert("Uh-oh, couldn't load Order");
+        this.router.navigate(['/']);
+        return;
+      }
+
+      const records = res.records[0] && res.records[0].data && res.records[0].data.searchResults;
+
+      this.data = records && records[0] && new OrderLNG(records[0]);
+
+      this.isLoading = false;
+    });
+  }
+
+  publish(): void {
+    this.isPublishing = true;
+
+    this.factoryService
+      .publishRecord(this.data)
+      .pipe(takeUntil(this.ngUnsubscribe))
+      .subscribe(response => {
+        this.data = new OrderLNG(response);
+
+        this.isPublishing = false;
+      });
+  }
+
+  unPublish(): void {
+    this.isUnPublishing = true;
+
+    this.factoryService
+      .unPublishRecord(this.data)
+      .pipe(takeUntil(this.ngUnsubscribe))
+      .subscribe(response => {
+        this.data = new OrderLNG(response);
+
+        this.isUnPublishing = false;
+      });
+  }
+
+  isPublished(): boolean {
```

Comment:
Instead of calling isPublished() on an *ngIf, this should be calculated/cached at ngOnInit() time or re-calculated and set to a variable right before doing change detection.  Same goes for other functions inside the *ngIf.

Reply:
Not 100% sure if they way I changed it is what you had in mind, let me know.

Reply:
This is better using a variable, however one last little bit is missing.  Here and in other places you implemented ngOnChanges() which is not quite what we've normally been doing.  That function is called when changes are changed in input based data bound variables.  Once you change data bound props inside the subscribe eg: `this.foo = 'bar'`, you need to tell angular to run the change detection cycle as it's possible that the service broke out of angular's zone.  Remove the ngOnChanges, and instead Import changeDetection and call the detectChanges() function on it, like so: `this._changeDetectionRef.detectChanges();`

See `inspection-detail.component.ts` in public-nrpti project for an example.

Reply:
Ah right, I forgot about that import.

Cool, fixed.

I also moved the status string to a variable, instead of a function referenced in the template.